src/NUnitFramework/framework/Internal/MessagePumpStrategy.cs

```diff
@@ -0,0 +1,139 @@
+// ***********************************************************************
+// Copyright (c) 2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Threading;
+
+#if NET20 || NET35 || NET40 || NET45
+using System.Windows.Forms;
+#if !NET20
+using System.Windows.Threading;
+#endif
+#endif
+
+namespace NUnit.Framework.Internal
+{
+    internal abstract class MessagePumpStrategy
+    {
+        public abstract void WaitForCompletion(AwaitAdapter awaitable);
+
+        public static MessagePumpStrategy FromCurrentSynchronizationContext()
+        {
+#if NET20 || NET35 || NET40 || NET45
```

Comment:
Once again, why the support for platforms that don't support async?

Reply:
Originally this file wasn't related to C#'s async. Now that I've changed it to receive an object so that the message pump strategy can decide whether to use a callback or to block, and I called the object AwaitAdapter, it really makes sense to make this only show up along with async support. I'll use `ASYNC`.

Reply:
Once I add `#if ASYNC` at the top, here I'll need to choose between doing `#if !(NETSTANDARD1_6 || NETSTANDARD2_0)` and `#if NET40 || NET45`.